src/udev/udev-builtin-path_id.c

```diff
@@ -635,8 +636,24 @@ static int find_real_nvme_parent(sd_device *dev, sd_device **ret) {
                 return r;
 
         /* If the 'real parent' is (still) virtual, e.g. for nvmf disks, refuse to set ID_PATH. */
```

Comment:
Please update the comment.

Reply:
@mwilck 
Thank you for pointing out that. Several questions.

> The code above (`strndupa_safe(sysname, end - sysname)`) assumes that the subsystem number in `/sys/class/nvme-subsystem/nvme-subsys${subsys}/nvme${subsys}n${ns}` translates 1:1 into a controller ID under `/sys/class/nvme/nvme${controller}`. This is problematic per se, as subsystem and controller IDs are independent, but not the topic of this PR (I suppose it works ok-ish for PCI, where subsystems correspond to controllers, at least without multipath).

- Looking at https://github.com/torvalds/linux/blob/b3603fcb79b1036acae10602bffc4855a4b9af80/drivers/nvme/host/core.c#L2903, it seems that the name of a subsystem is set by the controller, no? How the instances of controller and subsystem can differ?
- I am not familiar with multipath, what is the correspondence between subsys, controller, and namespace? Can a controller belongs to multiple subsystems? And/or can a subsystem can constructed with multiple controllers? How can we get the relation of them?

> But in the context of this PR, it means that you will create a symlink only for devices for which the subsystem ID is the same as the controller ID. In the multipath case, this will create only one `by-path` symlink to every namespace. If this path goes away, the symlink will go away too, even if in a multipaht scenario the namespace would still exist and would probably be accessible just fine through other paths.

If the main(?) path goes away, then does the kernel send 'change' (or any other) uevent? Then, we can recreate symlink to another by-path symlink, no?

Reply:
> Looking at https://github.com/torvalds/linux/blob/b3603fcb79b1036acae10602bffc4855a4b9af80/drivers/nvme/host/core.c#L2903, it seems that the name of a subsystem is set by the controller, no? How the instances of controller and subsystem can differ?

The code is confusing. A few lines below the kernel looks up the subsystem by NQN, and if it finds it, it will discard the just initialized `subsys`. You are right (and I admit I wasn't aware of that) that the initial ID of the subsystem is not simply enumerated, it is equal to ID of the controller where this subsystem was first attached. However, especially on NVMeoF with multipath, there are multiple contollers with different IDs. The first controller an later be disconnected, so that there is no match any more between the subsystem ID and the ID of any of the controllers, and this remains so even if the connection is re-established later.

Example below. First we see a match between subsys ID (14) and controller ID. Then I disconnect and re-connect the connection of `nvme14`. The new connection gets the controller ID 0.

```
# nvme list-subsys
nvme-subsys14 - NQN=<censored>
\
 +- nvme14 fc traddr=nn-0x200200a09890f5bf:pn-0x203800a09890f5bf,host_traddr=nn-0x20000024ff7fa448:pn-0x21000024ff7fa448 live 
 +- nvme16 fc traddr=nn-0x200200a09890f5bf:pn-0x203600a09890f5bf,host_traddr=nn-0x20000024ff7fa448:pn-0x21000024ff7fa448 live 

# nvme disconnect -d /dev/nvme14
# nvme list-subsys
nvme-subsys14 - NQN=<censored>
\
 +- nvme16 fc traddr=nn-0x200200a09890f5bf:pn-0x203600a09890f5bf,host_traddr=nn-0x20000024ff7fa448:pn-0x21000024ff7fa448 live 

# nvme connect -t fc -n <censored> -a nn-0x200200a09890f5bf:pn-0x203800a09890f5bf
# nvme list-subsys
nvme-subsys14 - NQN=<censored>
\
 +- nvme0 fc traddr=nn-0x200200a09890f5bf:pn-0x203800a09890f5bf,host_traddr=nn-0x20000024ff7fa448:pn-0x21000024ff7fa448 live 
 +- nvme16 fc traddr=nn-0x200200a09890f5bf:pn-0x203600a09890f5bf,host_traddr=nn-0x20000024ff7fa448:pn-0x21000024ff7fa448 live 

# (cd /sys/class/nvme-subsystem/nvme-subsys14 && ls -ld nvme*)
lrwxrwxrwx 1 root root 0 Mar 19 11:12 nvme0 -> ../../nvme-fabrics/ctl/nvme0
drwxr-xr-x 8 root root 0 Mar 13 09:55 nvme14n1
drwxr-xr-x 8 root root 0 Mar 13 09:55 nvme14n2
lrwxrwxrwx 1 root root 0 Mar 13 09:55 nvme16 -> ../../nvme-fabrics/ctl/nvme16
```

Reply:
> I am not familiar with multipath, what is the correspondence between subsys, controller, and namespace? Can a controller belongs to multiple subsystems? And/or can a subsystem can constructed with multiple controllers? How can we get the relation of them?

Picture a subsystem as a storage array with multiple storage units (aka namespaces)[^namespace] and multiple I/O controllers. The "nvme controller" represents actually the "path" between a host and the subsystem, e.g. the connection of a host FC port with an FC port of of the storage. In a typical multipath scenario, you have 2 ports on the storage connected to 2 ports on the host each, thus 4 NVMe controllers.

[^namespace]: in the simplest case, a namespace would be one SSD in the storage array (industry calls this "JBOD" - "just a bunch of 'disks'"). In more complex storage, it's a logical mapping of logical extents towards one ore more physical extents on real SSDs, similar to LVM.

Reply:
> If the main(?) path goes away, then does the kernel send 'change' (or any other) uevent? Then, we can recreate symlink to another by-path symlink, no?

AFAICS, the kernel sends only uevents for the NVMe controller. 
Here is an example for removal/addition like above:
```
KERNEL[524421.207624] remove   /devices/virtual/nvme-fabrics/ctl/nvme0/hwmon6 (hwmon)
KERNEL[524421.207740] remove   /devices/virtual/nvme-fabrics/ctl/nvme0 (nvme)
KERNEL[524455.529493] add      /devices/virtual/nvme-fabrics/ctl/nvme0 (nvme)
KERNEL[524456.749687] add      /devices/virtual/nvme-fabrics/ctl/nvme0/hwmon6 (hwmon)
KERNEL[524457.249886] change   /devices/virtual/nvme-fabrics/ctl/nvme0 (nvme)
```

Substituting the symlink would be quite difficult. udev would need to figure out the subsystem which was associated with the controller being removed (not sure if the links in sysfs still exist when the remove uevent arrives!), and create a symlink for a *different* path device.

I don't think we do anything like this for any other subsystem. The idea sounds quite fragile to me.

Reply:
When a new path is added for an existing subsystem like above, we could check if a by-path link for the subsystem exists, and add one if not. But this would obviously not be atomic. The add event may come minutes later, or not at all.

Reply:
Note that if the `by-path` symlink was used for mounting the device, even if we could substitute a new, different `by-path` link for the same subsystem, the original `by-path` link would be gone and system would consider the respective device "dead", and possibly unmount the file system.

Reply:
OTOH, if the symlink is _not_ used for mounting, we just don't need it.

The more I think about it, the more I come to the conclusion that adding these symlinks creates more harm than benefit if multipath is involved. 

If `nvme_core.multipath=Y` (the default), we'd need to check the namespace's "nmic" field of the "identify namespace" command (NVMe base spec Fig. 281, `nvme id-ns -o json /dev/nvme${X}n${Y} | jq .nmic`). If NMIC is zero, the namespace can only be attached to a single controller, and creating a `by-path` ID would be safe, but otherwise it would be affected by the stability issues noted above.

Reply:
FWIW, I don't think it's a big problem to add `NMIC` to sysfs if needed.

Reply:
Udev code can't rely on that, though. It must work with older kernels, too.

Reply:
Strange rule, but well, if you don't want shiny things, that is also okay for me :)

Reply:
Below is example for multipath nvme device(2 paths):

# nvme list
/dev/nvme1n1          /dev/ng1n1            S5H0NE0MC00161       3.2TB NVMe Gen4

# nvme list-subsys
nvme-subsys1 - NQN=nqn.1994-11.com.samsung:nvme:PM1735:2.5-inch:S5H0NE0MC00161
\
 +- nvme0 pcie 0536:b8:00.0 live
 +- nvme1 pcie 0596:b8:00.0 live

# lspci
0536:b8:00.0 Non-Volatile memory controller: Samsung Electronics Co Ltd NVMe SSD Controller PM173X
0596:b8:00.0 Non-Volatile memory controller: Samsung Electronics Co Ltd NVMe SSD Controller PM173X
#31873 

# ls -l /dev/disk/by-path
total 0
lrwxrwxrwx. 1 root root 13 Mar 20 13:05 pci-0596:b8:00.0-nvme-1 -> ../../nvme1n1

[root@ltcrain119-lp8 ~]# ls -l /dev/disk/by-id/nvme*
lrwxrwxrwx. 1 root root 13 Mar 20 13:05 /dev/disk/by-id/nvme-3.2TB_NVMe_Gen4_U.2_SSD_S5H0NE0MC00161 -> ../../nvme1n1
lrwxrwxrwx. 1 root root 13 Mar 20 13:05 /dev/disk/by-id/nvme-3.2TB_NVMe_Gen4_U.2_SSD_S5H0NE0MC00161_1 -> ../../nvme1n1
lrwxrwxrwx. 1 root root 13 Mar 20 13:05 /dev/disk/by-id/nvme-eui.354830304dc001610025384500000001 -> ../../nvme1n1

Now removing nvme1->pci 0596

# lspci
0536:b8:00.0 Non-Volatile memory controller: Samsung Electronics Co Ltd NVMe SSD Controller PM173X
# nvme list
Node                  Generic               SN                   Model                                    Namespace  Usage                      Format           FW Rev
--------------------- --------------------- -------------------- ---------------------------------------- ---------- -------------------------- ---------------- --------
/dev/nvme1n1          /dev/ng1n1            S5H0NE0MC00161       3.2TB NVMe Gen4 U.2 SSD                  0x1        343.60  GB / 343.60  GB      4 KiB +  0 B   REV.ST45
# nvme list-subsys
nvme-subsys1 - NQN=nqn.1994-11.com.samsung:nvme:PM1735:2.5-inch:S5H0NE0MC00161
\
 +- nvme0 pcie 0536:b8:00.0 live
# ls -l /dev/disk/by-path/
total 0
lrwxrwxrwx. 1 root root 13 Mar 20 13:05 pci-0596:b8:00.0-nvme-1 -> ../../nvme1n1
# ls -l /dev/disk/by-id/nvme*
lrwxrwxrwx. 1 root root 13 Mar 20 13:05 /dev/disk/by-id/nvme-3.2TB_NVMe_Gen4_U.2_SSD_S5H0NE0MC00161 -> ../../nvme1n1
lrwxrwxrwx. 1 root root 13 Mar 20 13:05 /dev/disk/by-id/nvme-3.2TB_NVMe_Gen4_U.2_SSD_S5H0NE0MC00161_1 -> ../../nvme1n1
lrwxrwxrwx. 1 root root 13 Mar 20 13:05 /dev/disk/by-id/nvme-eui.354830304dc001610025384500000001

Still keep /dev/nvme1n1 in by-path and by-id, this is designed by nvme community. traffic io over nvme1n1 don't need to restart.

Then adding back pci 0596,

]# lspci
0536:b8:00.0 Non-Volatile memory controller: Samsung Electronics Co Ltd NVMe SSD Controller PM173X
0596:b8:00.0 Non-Volatile memory controller: Samsung Electronics Co Ltd NVMe SSD Controller PM173X
# nvme list
Node                  Generic               SN                   Model                                    Namespace  Usage                      Format           FW Rev
--------------------- --------------------- -------------------- ---------------------------------------- ---------- -------------------------- ---------------- --------
/dev/nvme1n1          /dev/ng1n1            S5H0NE0MC00161       3.2TB NVMe Gen4 U.2 SSD                  0x1        343.60  GB / 343.60  GB      4 KiB +  0 B   REV.ST45
# nvme list-subsys
nvme-subsys1 - NQN=nqn.1994-11.com.samsung:nvme:PM1735:2.5-inch:S5H0NE0MC00161
\
 +- nvme0 pcie 0536:b8:00.0 live
 +- nvme2 pcie 0596:b8:00.0 live
# ls -l /dev/disk/by-path
total 0
lrwxrwxrwx. 1 root root 13 Mar 20 13:05 pci-0596:b8:00.0-nvme-1 -> ../../nvme1n1
# ls -l /dev/disk/by-id/nvme*
lrwxrwxrwx. 1 root root 13 Mar 20 13:05 /dev/disk/by-id/nvme-3.2TB_NVMe_Gen4_U.2_SSD_S5H0NE0MC00161 -> ../../nvme1n1
lrwxrwxrwx. 1 root root 13 Mar 20 13:05 /dev/disk/by-id/nvme-3.2TB_NVMe_Gen4_U.2_SSD_S5H0NE0MC00161_1 -> ../../nvme1n1
lrwxrwxrwx. 1 root root 13 Mar 20 13:05 /dev/disk/by-id/nvme-eui.354830304dc001610025384500000001

Still keep nvme1n1 in by-path and by-id.

Above is an example for 2 paths nvme device.